### PR TITLE
user error trigger

### PR DESCRIPTION
**Pull request by tiendt**
Milestone 1: end-to-end function

Tasks:
Introduce new trigger type “Error trigger”. (`events.js`)
Accept the new error trigger from config. (`instrumentation.js`)
Intercept user logger and trigger amp-analytics on errors (`log.js`)

### CODE REVIEW

File: examples/analytics-error.amp.html [hunk 1]
```diff
@@ -0,0 +1,36 @@
+<!doctype html>
+<html amp lang="en">
+<head>
+    <meta charset="utf-8">
+    <script async src="https://cdn.ampproject.org/v0.js"></script>
+    <title>amp-analytics</title>
+    <link rel="canonical" href="http://example.ampproject.org/article-metadata.html" />
+    <meta name="viewport" content="width=device-width,minimum-scale=1,initial-scale=1">
+    <script type="application/ld+json">
```
Comment (lannka) [inline examples/analytics-error.amp.html]: this is unreleated, you can remove

File: examples/analytics-error.amp.html [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+        "datePublished": "2015-10-07T12:02:41Z",
+        "image": [
+          "logo.jpg"
+        ]
+      }
+    </script>
+    <style amp-boilerplate>body{-webkit-animation:-amp-start 8s steps(1,end) 0s 1 normal both;-moz-animation:-amp-start 8s steps(1,end) 0s 1 normal both;-ms-animation:-amp-start 8s steps(1,end) 0s 1 normal both;animation:-amp-start 8s steps(1,end) 0s 1 normal both}@-webkit-keyframes -amp-start{from{visibility:hidden}to{visibility:visible}}@-moz-keyframes -amp-start{from{visibility:hidden}to{visibility:visible}}@-ms-keyframes -amp-start{from{visibility:hidden}to{visibility:visible}}@-o-keyframes -amp-start{from{visibility:hidden}to{visibility:visible}}@keyframes -amp-start{from{visibility:hidden}to{visibility:visible}}</style><noscript><style amp-boilerplate>body{-webkit-animation:none;-moz-animation:none;-ms-animation:none;animation:none}</style></noscript>
+</head>
+<body>
+<h1>amp-analytics error tracker tester</h1>
```
Comment (lannka) [inline examples/analytics-error.amp.html]: is it empty now? you can include it in later PR

File: extensions/amp-analytics/0.1/error-impl.js [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {registerServiceBuilder, getService} from '../service';
+import {triggerAnalyticsEvent} from '../../../src/analytics';
+
+export class ErrorService {
```
Comment (lannka) [inline extensions/amp-analytics/0.1/error-impl.js]: UserErrorReportingService

File: extensions/amp-analytics/0.1/error-impl.js [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+    //put error to queue????
+    //receive error
```
Comment (lannka) [inline extensions/amp-analytics/0.1/error-impl.js]: use format: `//TODO(tiendt, #9821): bla bla`

File: extensions/amp-analytics/0.1/error-impl.js [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+ */
+export function installUserErrorService(window) {
+  registerServiceBuilder(window, 'error', ErrorService);
+}
+
+/**
+ * @param {!Window} window
+ * @return {!ErrorService}
+ */
+export function userErrorFor(window) {
```
Comment (lannka) [inline extensions/amp-analytics/0.1/error-impl.js]: this will go into `/src/services/`

File: extensions/amp-analytics/0.1/events.js [hunk 6]
```diff
@@ -507,6 +507,34 @@ export class VisibilityTracker extends EventTracker {
   }
 }
 
+export class ErrorTracker extends EventTracker {
```
Comment (lannka) [inline extensions/amp-analytics/0.1/events.js]: i thought you're thinking of recusing `CustomEventTracker` ? What's the final decision?

File: src/log.js [hunk 7]
```diff
@@ -17,6 +17,9 @@
 import {getMode} from './mode';
 import {getModeObject} from './mode-object';
 import {isEnumValue} from './types';
+import {isExperimentOn} from '../../../src/experiments';
+import {installUserErrorService} from '../extensions/amp-analytics/0.1/error-impl';
```
Comment (lannka) [inline src/log.js]: don't import here

File: build-system/global-configs/canary-config.json [hunk 8]
```diff
@@ -25,5 +25,6 @@
   "amp-bind": 1,
   "ad-loader-v2": 1,
   "3p-use-ampcontext": 1,
+  "user-error-tracker": 0
```
Comment (lannka) [inline build-system/global-configs/canary-config.json]: no need to change, if it's 0

File: extensions/amp-analytics/0.1/error-impl.js (hunk 3 above)
Comment (lannka) [inline extensions/amp-analytics/0.1/error-impl.js]: I would see the public methods in this PR (skeleton is fine).

File: src/log.js [hunk 9]
```diff
@@ -583,3 +586,21 @@ export function dev() {
     return LogLevel.OFF;
   });
 }
+
+
+/**
+ * @param error
+ */
+export function sendErrorToService(error) {
```
Comment (lannka) [inline src/log.js]: please add test for this

**Comment by lannka** [comment]
Please also update PR title and description

File: src/log.js [hunk 10]
```diff
@@ -201,77 +206,105 @@ export class Log {
...
-  /**
-   * Reports an error message and marks with an expected property. If the
-   * logging is disabled, the error is rethrown asynchronously.
-   * @param {string} unusedTag
-   * @param {...*} var_args
-   */
+    /**
+     * @param {!Error} error
+     */
+  sendErrorToService(error) {
```
Comment (lannka) [inline src/log.js]: reportErrorToAnalytics

File: src/log.js [hunk 11]
```diff
@@ -201,77 +206,105 @@ export class Log {
...
+    if (isExperimentOn(this.win, 'user-error-tracker')) {
+            /**
+             * @param {!Window} win
+             */
+      const vars = {
+        'errorName': error.name,
```
Comment (lannka) [inline src/log.js]: make it tag

File: src/log.js [hunk 12]
```diff
@@ -201,77 +206,105 @@ export class Log {
...
+        'errorMessage': error.message,
+      };
+      triggerAnalyticsEvent(this.getRootElement_(),'user-error-trigger', vars);
```
Comment (lannka) [inline src/log.js]: user-error

File: src/log.js [hunk 13]
```diff
@@ -201,77 +206,105 @@ export class Log {
...
-   * Reports an error message and marks with an expected property. If the
-   * logging is disabled, the error is rethrown asynchronously.
-   * @param {string} unusedTag
-   * @param {...*} var_args
-   */
+    /**
+     * @param {!Error} error
+     */
+  sendErrorToService(error) {
+    if (isExperimentOn(this.win, 'user-error-tracker')) {
```
Comment (lannka) [inline src/log.js]: user-error-reporting

File: src/log.js [hunk 14]
```diff
@@ -69,63 +71,66 @@ export function setReportError(fn) {
...
-   * @param {!Window} win
-   * @param {function(!./mode.ModeDef):!LogLevel} levelFunc
-   * @param {string=} opt_suffix
-   */
-  constructor(win, levelFunc, opt_suffix) {
     /**
-     * In tests we use the main test window instead of the iframe where
-     * the tests runs because only the former is relayed to the console.
-     * @const {!Window}
+     * @param {!Window} win
```
Comment (lannka) [inline src/log.js]: revert all the indentation

File: examples/analytics-error.amp.html [hunk 15]
```diff
@@ -0,0 +1,56 @@
+<!doctype html>
```
Comment (lannka) [inline examples/analytics-error.amp.html]: analytics-error-reporting

File: examples/analytics-error.amp.html [hunk 16]
```diff
@@ -0,0 +1,56 @@
...
+<html ⚡>
+<head>
+    <meta charset="utf-8">
+    <title>Lorem Ipsum</title>
+    <link rel="canonical" href="amps.html" >
+    <meta name="viewport" content="width=device-width,minimum-scale=1,initial-scale=1">
+    <link href='https://fonts.googleapis.com/css?family=Questrial' rel='stylesheet' type='text/css'>
```
Comment (lannka) [inline examples/analytics-error.amp.html]: not needed

File: examples/analytics-error.amp.html [hunk 17]
```diff
@@ -0,0 +1,56 @@
...
+    Curabitur ullamcorper turpis vel commodo scelerisque. Phasellus
+    luctus nunc ut elit cursus, et imperdiet diam vehicula.
+    Duis et nisi sed urna blandit bibendum et sit amet erat.
+</p>
+
+<amp-analytics>
+    <script type="application/json">
+        {
+            "requests": {
+                "user-error-trigger": "http://localhost:8000/dsdsdsd"
```
Comment (lannka) [inline examples/analytics-error.amp.html]: update name here

File: src/log.js [hunk 18]
```diff
@@ -212,10 +217,39 @@ export class Log {
...
+
+  /**
+   * @param {string} unusedTag
+   * @param {!Error} error
+   */
+  reportErrorToAnalytics(unusedTag, error) {
+    if (isExperimentOn(this.win, 'user-error-reporting')) {
+      /**
+       * @param {!Window} win
+       */
```
Comment (lannka) [inline src/log.js]: remove this doc

File: src/log.js [hunk 19]
```diff
@@ -90,6 +92,9 @@ export class Log {
 
     /** @private @const {string} */
     this.suffix_ = opt_suffix || '';
+
+    /** @private @const {boolean} */
+    this.isUserError_ = !!opt_suffix;
```
Comment (lannka) [inline src/log.js]: check explicitly against `USER_ERROR_SENTINEL`

File: src/log.js [hunk 20]
```diff
@@ -212,10 +217,39 @@ export class Log {
     if (error) {
       // reportError is installed globally per window in the entry point.
       self.reportError(error);
+      if (this.isUserError_) {
+        this.reportErrorToAnalytics(unusedTag, error);
```
Comment (lannka) [inline src/log.js]: semantically, I think we should move this into `error.js reportError()`

File: build-system/tasks/presubmit-checks.js [hunk 21]
```diff
@@ -15,37 +15,37 @@
  */
 'use strict';
 
-var gulp = require('gulp-help')(require('gulp'));
-var path = require('path');
-var srcGlobs = require('../config').presubmitGlobs;
-var util = require('gulp-util');
-var through2 = require('through2');
+const gulp = require('gulp-help')(require('gulp'));
```
Comment (lannka) [inline build-system/tasks/presubmit-checks.js]: revert the var -> const change in this PR.
(it can be a separate PR)

File: build-system/tasks/presubmit-checks.js [hunk 22]
```diff
@@ -340,6 +340,7 @@ var forbiddenTerms = {
       'extensions/amp-access/0.1/amp-access.js',
       'extensions/amp-experiment/0.1/variant.js',
       'extensions/amp-user-notification/0.1/amp-user-notification.js',
+      'src/log.js',
```
Comment (lannka) [inline build-system/tasks/presubmit-checks.js]: should be the `dist.3p/current/integration.js`

File: build-system/tasks/presubmit-checks.js [hunk 23]
```diff
@@ -1003,7 +1003,7 @@ function checkForbiddenAndRequiredTerms() {
       if (missingRequirements) {
         util.log(util.colors.blue(
             'Adding these terms (e.g. by adding a required LICENSE ' +
-            'to the file)'));
+          'to the file)'));
```
Comment (lannka) [inline build-system/tasks/presubmit-checks.js]: revert this

File: test/functional/test-error.js [hunk 24]
```diff
@@ -539,3 +541,32 @@ describe('detectJsEngineFromStack', () => {
...
+
+describes.fakeWin('user error reporting', {amp: true}, env => {
+  let win;
+  adopt(window);
+  const error = new Error('ERROR','user error');
+  let analyticsEventSpy_;
+
+  beforeEach(() => {
+    win = env.win;
+    sandbox = sinon.sandbox.create();
```
Comment (lannka) [inline test/functional/test-error.js]: sandbox = env.sandbox;

also declare it below `let win`;

File: test/functional/test-error.js [hunk 25]
```diff
@@ -539,3 +541,32 @@ describe('detectJsEngineFromStack', () => {
     });
   });
 });
+
+
+describes.fakeWin('user error reporting', {amp: true}, env => {
+  let win;
+  adopt(window);
```
Comment (lannka) [inline test/functional/test-error.js]: no need to call this

File: test/functional/test-error.js [hunk 26]
```diff
@@ -539,3 +541,32 @@ describe('detectJsEngineFromStack', () => {
...
+
+  beforeEach(() => {
+    win = env.win;
+    sandbox = sinon.sandbox.create();
+    analyticsEventSpy_ = sandbox.spy(analytics, 'triggerAnalyticsEvent');
+    toggleExperiment(window, 'user-error-reporting', true);
+  });
+
+  afterEach(() => {
+    sandbox.restore();
```
Comment (lannka) [inline test/functional/test-error.js]: no need to do, already handled by describes

**Comment by tiendt** [comment]
@lannka I updated the tests and all

File: src/log.js [hunk 27]
```diff
@@ -209,6 +209,7 @@ export class Log {
   error(unusedTag, var_args) {
     const error = this.error_.apply(this, arguments);
     if (error) {
+      error.name = unusedTag || error.name;
```
Comment (lannka) [inline src/log.js]: rename unusedTag -> tag

File: test/functional/test-error.js [hunk 28]
```diff
@@ -539,3 +540,26 @@ describe('detectJsEngineFromStack', () => {
     });
   });
 });
+
+
+describes.fakeWin('user error reporting', {amp: true}, env => {
+  let win;
+  sandbox = env.sandbox;
+  const error = new Error('ERROR','user error');
+  let analyticsEventSpy_;
```
Comment (lannka) [inline test/functional/test-error.js]: no need for trailing `_`. that naming convention only apply for class members.
